pkg/applicationserver/applicationserver.go

```diff
@@ -450,30 +449,70 @@ func (as *ApplicationServer) buildSessionsFromError(ctx context.Context, dev *tt
 
 const maxDownlinkQueueOperationAttempts = 50
 
-func (as *ApplicationServer) downlinkQueueOp(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, items []*ttnpb.ApplicationDownlink, op func(ttnpb.AsNsClient, context.Context, *ttnpb.DownlinkQueueRequest, ...grpc.CallOption) (*pbtypes.Empty, error)) error {
-	ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("as:downlink:%s", events.NewCorrelationID()))
-	for _, item := range items {
-		item.CorrelationIDs = append(item.CorrelationIDs, events.CorrelationIDsFromContext(ctx)...)
-	}
-	logger := log.FromContext(ctx)
-	for _, item := range items {
-		registerReceiveDownlink(ctx, ids, item)
-	}
-	peer, err := as.GetPeer(ctx, ttnpb.ClusterRole_NETWORK_SERVER, &ids)
-	if err != nil {
-		return err
-	}
+func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, dev *ttnpb.EndDevice, items []*ttnpb.ApplicationDownlink, op func(ttnpb.AsNsClient, context.Context, *ttnpb.DownlinkQueueRequest, ...grpc.CallOption) (*pbtypes.Empty, error), link *ttnpb.ApplicationLink, peer cluster.Peer) ([]string, error) {
 	pc, err := peer.Conn()
 	if err != nil {
-		return err
+		return nil, err
 	}
+	mask := make([]string, 0, 2)
+	attempt := 0
+	for {
+		ctx := log.NewContextWithField(ctx, "attempt", attempt)
+
+		sessions := make([]*ttnpb.Session, 0, 2)
+		if dev.Session != nil {
+			sessions = append(sessions, dev.Session)
+			mask = append(mask, "session.last_a_f_cnt_down")
+		}
+		if dev.PendingSession != nil {
+			// Downlink can be encrypted with the pending session while the device first joined but not confirmed the
+			// session by sending an uplink.
+			sessions = append(sessions, dev.PendingSession)
+			mask = append(mask, "pending_session.last_a_f_cnt_down")
+		}
+		if len(sessions) == 0 {
+			return nil, errNoDeviceSession.New()
+		}
+
+		encryptedItems, err := as.encodeAndEncryptDownlinks(ctx, dev, link, items, sessions)
+		if err != nil {
+			return nil, err
+		}
+		_, err = op(ttnpb.NewAsNsClient(pc), ctx, &ttnpb.DownlinkQueueRequest{
+			EndDeviceIdentifiers: ids,
+			Downlinks:            encryptedItems,
+		}, as.WithClusterAuth())
+		if err == nil {
+			return mask, nil
+		}
+		if attempt >= maxDownlinkQueueOperationAttempts || as.skipPayloadCrypto(ctx, link, dev, nil) {
```

Comment:
First attempt has `attempt == 0`, while `maxDownlinkQueueOperationAttempts` is a counter. So having that value set to `1` results in two attempts.